File: 2025-06-20/AsteroidCollision.java
import java.util.*;

public class AsteroidCollision {
    public static int[] asteroidCollision(int[] asteroids) {
        if(asteroids == null || asteroids.length == 0) return new int[]{};

        Stack<Integer> stack = new Stack<>();
        for(int ast : asteroids) {
            boolean flag = true;
            while(!stack.isEmpty() && flag && stack.peek() > 0 && ast < 0) {
                int diff = Math.abs(stack.peek()) - Math.abs(ast);
                if(diff == 0) {
                    stack.pop();
                    flag = false;
                }
                else if(diff > 0) flag = false;
                else stack.pop();
            }
            if(flag)
                stack.add(ast);
        }

        int[] result = new int[stack.size()];
        for(int i=result.length-1;i>=0;i--)
            result[i] = stack.pop();

        return result;
    }
    public static void main(String[] args) {
        int[] arr = {-1,-1,-1,10,-120};
        System.out.println(Arrays.toString(asteroidCollision(arr)));
    }
}

File: 2026-01-28/Asteroid_Collision.py
from typing import List


class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        if not asteroids:
            return []

        stack = []
        for asteroid in asteroids:
            while stack and asteroid < 0 < stack[-1]:
                collision = stack[-1] + asteroid
                if collision < 0:
                    stack.pop()
                    continue
                elif collision == 0:
                    stack.pop()
                break
            else:
                stack.append(asteroid)

        return stack


sol = Solution()
print(sol.asteroidCollision([5,10,-5]))